### PR TITLE
Add scatter/contour/contourf functionality

### PR DESCRIPTION
**Pull request by briochemc**
This may be a bit dodgy but it seems to work on the base examples that I added to the readme... Have a look and let me know what you think!

### CODE REVIEW

File: src/UnitfulRecipes.jl [hunk 1]
```diff
@@ -42,11 +47,15 @@ function resolve_axis!(attr, arr::AbstractArray{T}, axis::Int) where {T<:Quantit
 
     # get axis label and append unit
     key = key_label(axis)
+    ckey = (axis==3) ? :colorbar_title : key
     if haskey(attr, key)
         attr[key] = string(attr[key], " ($(string(u)))")
+        attr[ckey] = attr[key]
     else
         attr[key] = string(u)
+        attr[ckey] = attr[key]
```
Comment (jw3126) [inline src/UnitfulRecipes.jl]: You do this on both branches of the if?

File: src/UnitfulRecipes.jl [hunk 2]
```diff
@@ -12,6 +12,11 @@ function recipe!(attr, arr)
 end
 
 function recipe!(attr, arrs...)
+    if haskey(attr, :marker_z)
+        u = unit(eltype(attr[:marker_z]))
+        attr[:marker_z] = ustrip.(u, attr[:marker_z])
+        attr[:colorbar_title] = string(u)
```
Comment (jw3126) [inline src/UnitfulRecipes.jl:32]: What does this do if `marker_z` is unitless?

File: src/UnitfulRecipes.jl (hunk 1 above)
Comment (briochemc) [reply src/UnitfulRecipes.jl]: Yes apologies this is a bit dirty... I just wanted a working prototype for `contourf` originally... I would need more time to make it neatly :/

File: src/UnitfulRecipes.jl (hunk 2 above)
Comment (briochemc) [reply src/UnitfulRecipes.jl:32]: I guess it leaves it as is? Again sorry this is dirty...

**Comment by jw3126** [comment]
With the README becoming longer, we can try to test it. See for inspiration:
* https://github.com/tkf/Kaleido.jl/blob/a209a0c80ee5ed18ee758021f50e8a0441e8fed9/src/Kaleido.jl#L4
* https://github.com/jw3126/Setfield.jl/blob/master/test/runtests.jl#L18

File: src/UnitfulRecipes.jl (hunk 2 above)
Comment (jw3126) [reply src/UnitfulRecipes.jl:32]: It's a honest question, I guess `u == NoUnits()` and becomes an empty string?

**Comment by jw3126** [comment]
Another possibility might be to dump all the test plots for inspection. Maybe something like this:
```julia
julia> using Test

julia> TEST_PLOT_DIR = tempname()
"/tmp/jl_YZTBeB"

julia> PLOT_COUNTER = 0
0

julia> @info "TEST_PLOT_DIR = $TEST_PLOT_DIR"
[ Info: TEST_PLOT_DIR = /tmp/jl_YZTBeB

julia> macro testplot(call)
           global PLOT_COUNTER += 1
           path = joinpath(TEST_PLOT_DIR, string(PLOT_COUNTER, call,  ".png"))
           esc(quote
               plt = $call
               @test plt isa Plots.Plot
               
               png($path)
           end)
       end
@testplot (macro with 1 method)

julia> @macroexpand1 @testplot plot(x,y, xlims=(1m, 2m))
quote
    #= REPL[5]:5 =#
    plt = plot(x, y, xlims=(1m, 2m))
    #= REPL[5]:6 =#
    #= REPL[5]:6 =# @test plt isa Plots.Plot
    #= REPL[5]:8 =#
    png("/tmp/jl_YZTBeB/1plot(x, y, xlims=(1m, 2m)).png")
end
```

File: src/UnitfulRecipes.jl (hunk 2 above)
Comment (briochemc) [reply src/UnitfulRecipes.jl:32]: Yes I just checked and that's what happens. (sorry for the delay I was sleeping — Australian time here.)

**Comment by briochemc** [comment]
Or generate some docs with these examples? For example I did this [here](https://briochemc.github.io/AIBECS.jl/stable/howtos/2_plot/) using Literate.jl.

**Comment by jw3126** [comment]
Yes, the literate doces look great! It would be cool to have that here as well. In addition I think having a `@testplot` macro that expands to something like:
```julia
@macroexpand1 @testplot plot(x,y, xunit=m)
```
```julia
path = joinpath(@__DIR__, "regressiondata",  testsetname, "plot(x,y, xunit=m).png")
plt = plot(x,y, xunit=m)
if ispath(path)
    ref = load(path)
else
    ref = plt
    @info "$path does not exist, creating"
    save(path, ref)
end

@test plt is Plot
if !isapprox(ref, plt)
     @error "Test fail $path"
     display(plot(ref, plt))
end
@test isapprox(ref, plt)
```

**Comment by jw3126** [comment]
But yeah, all of these ideas do not have to be part of this PR.

**Comment by briochemc** [comment]
OK I have written a bunch of docs and examples and it sort of works locally but I'm unsure how I should merge them here... I could add them to this PR (this is OK with me because I personally do not care that much about the state of the master branch or the history of commits, as long as I can retrieve any tagged release). Maybe we can go ahead with that and then work from that merged?

**Comment by briochemc** [comment]
I'm not sure there is a way for you to see the docs other than either (i) you cloning my fork somewhere and deploying the docs locally or (ii) merging in your master and going through Documenter.jl's setup for hosting documentation...

**Comment by briochemc** [comment]
FWIW below are some screenshots of what the docs looks like for me locally:

Screen Shot 2020-01-22 at 2 16 24 PM
Screen Shot 2020-01-22 at 2 16 17 PM
Screen Shot 2020-01-22 at 2 16 12 PM
Screen Shot 2020-01-22 at 2 16 07 PM
Screen Shot 2020-01-22 at 2 15 49 PM

**Comment by briochemc** [comment]
BTW for docs to be deployed and hosted you must go through a couple of steps as instructed in Documenter.jl's [documentation](https://juliadocs.github.io/Documenter.jl/stable/man/hosting/).

**Comment by jw3126** [comment]
This looks amazing! Big thanks! I agree it is pragmatic to just merge this and then fiddle with the deployment afterwards.

File: docs/make.jl [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    nb && Literate.notebook(ipath, notebooks, execute = execute)
+end
+
+# Documentation structure
+ismd(f) = splitext(f)[2] == ".md" 
+pages(folder) = [joinpath(folder, f) for f in readdir(joinpath(src, folder)) if ismd(f)]
+
+makedocs(
+    sitename="UnitfulRecipes.jl",
+    doctest = false, # TODO guessing I should remove that when actually deploying?
```
Comment (jw3126) [inline docs/make.jl:24]: What does a doctest do with a `Plot` output?

File: src/UnitfulRecipes.jl [hunk 4]
```diff
@@ -42,32 +80,45 @@ function resolve_axis!(attr, arr::AbstractArray{T}, axis::Int) where {T<:Quantit
...
     # get axis label and append unit
     key = key_label(axis)
-    if haskey(attr, key)
-        attr[key] = string(attr[key], " ($(string(u)))")
-    else
-        attr[key] = string(u)
+    ckey = (axis==3) ? :colorbar_title : key
+    ustr = string(u)
+    if haskey(attr, key) # Append unit (only once) if axis label exists
+        i = findlast(ustr, attr[key])
```
Comment (jw3126) [inline src/UnitfulRecipes.jl:86]: Why not use `occursin(ustr, attr[key])`?

**Comment by jw3126** [comment]
I think there are a couple of things that could be cleaned up here, but I will merge and try to deploy anyway.

**Comment by jw3126** [comment]
@briochemc can you help:
https://travis-ci.com/jw3126/UnitfulRecipes.jl/jobs/278296133

**Comment by jw3126** [comment]
https://jw3126.github.io/UnitfulRecipes.jl/dev/examples/1_Examples/#Simplest-plot-1

sadly a lot of units are cut away in the docs. Any idea how to fix that @briochemc

File: docs/make.jl (hunk 3 above)
Comment (briochemc) [reply docs/make.jl:24]: I don’t know But I don’t think these are turned into doctests by Literate?

File: src/UnitfulRecipes.jl (hunk 4 above)
Comment (briochemc) [reply src/UnitfulRecipes.jl:86]: E.g. in case you have data in meters and you label it with a string that contains an "m".

**Comment by briochemc** [comment]
> @briochemc can you help:
> https://travis-ci.com/jw3126/UnitfulRecipes.jl/jobs/278296133

I'm not sure what you are pointing at here...

**Comment by briochemc** [comment]
> https://jw3126.github.io/UnitfulRecipes.jl/dev/examples/1_Examples/#Simplest-plot-1
> 
> sadly a lot of units are cut away in the docs. Any idea how to fix that @briochemc

Arf that's annoying... @mortenpi Maybe you know why plots in the deployed docs are cropped?

---

Edit: Actually I see them also cropped in the notebooks... So maybe a Plots.jl issue?

**Comment by briochemc** [comment]
FWIW I added a Project file in the gh-pages branch so the binder badges work now, and as I just tried one the plots looked alright... So I'm now unsure where the cropping issue comes from...

**Comment by jw3126** [comment]
> > @briochemc can you help:
> > https://travis-ci.com/jw3126/UnitfulRecipes.jl/jobs/278296133
> 
> I'm not sure what you are pointing at here...

In the log there is a lot of 
```
GKS: Open failed in routine OPEN_WS

GKS: GKS not in proper state. GKS must be either in the state WSOP or WSAC in routine ACTIVATE_WS

/home/travis/.julia/packages/GR/oiZD3/src/../deps/gr/bin/gksqt: error while loading shared libraries: libQt5Widgets.so.5: cannot open shared object file: No such file or directory

connect: Connection refused

GKS: can't connect to GKS socket application

Did you start 'gksqt'?
```
Do you know what it is about and whether we should care?

**Comment by briochemc** [comment]
Oh yes I asked before about that somewhere and I was told I should not care. I agree it looks scary and confusing though. IIRC there was a way to turn these warnings off, which I think is just adding [this line](https://github.com/JuliaPlots/Plots.jl/blob/3459bcd81766f71ba6ff0d185bf9e32cc5425521/test/runtests.jl#L50) in our tests.

**Pull request merged**